### PR TITLE
Pebble files API client functions

### PR DESCRIPTION
**Pull request by benhoyt**
This PR adds Python Operator Framework support for the new Pebble files API (https://github.com/canonical/pebble/pull/25). Two things to note:

* Per the [spec](https://docs.google.com/document/d/1NPPZZB_qCwtTru-YsKkRSAb6mXJzBiE0fG3GTZnx4fM/edit#), for now we're only supporting single-file operations in the client APIs, even though the HTTP API itself supports bulk operations.
* In addition, Python's standard library doesn't have good support for streaming writing/parsing of multipart, so the current code reads the whole file into memory. However, the API is such that we can fix this at a later date without changing the `read_file` and `write_file` signatures.

### CODE REVIEW

File: ops/pebble.py
Comment (wallyworld) [inline ops/pebble.py:600]: Create new FileInfo object from dict...    ?

File: ops/pebble.py
Comment (wallyworld) [inline ops/pebble.py:611]: Do we need to be defensive in case 'last-modified' is malformed?
Perhaps _parse_timestamp() handles it already

File: ops/pebble.py
Comment (wallyworld) [inline ops/pebble.py:676]: "expected" reads to me like a string literal value.
Perhaps `<expected>` ?

File: ops/pebble.py
Comment (wallyworld) [inline ops/pebble.py:900]: Where does 32 come from? Is that part of the spec? Or it's a reasonable minimum to test against for a coarse grained check?

**Review by wallyworld — COMMENTED** [review]
Quick initial pass through, looks nice

File: ops/pebble.py
Comment (jameinel) [inline ops/pebble.py:867]: It feels odd that this is
`read_file(path: str, destination: BinaryIO)`
rather than
`read_file(path: str) => BinaryIO`
I realize this somewhat exists to support bulk requests, but it does feel awkward.

I also wonder about a convenience function:
`def read_content(self, path: str) => bytes?/str?`

But that could be written on top of any sort of file based api.

File: ops/pebble.py
Comment (jameinel) [inline ops/pebble.py:943]: I think the general doc guideline is google docstrings with:
```
Args:
  user: user id or username for the owner of the file
  group: group id or group name for the file's group
...
```

I wouldn't block this PR on that, but we should be consistent about docstrings so that https://ops.rtfd.io works correctly

File: test/test_pebble.py
Comment (jameinel) [inline test/test_pebble.py:1331]: We don't seem to have a test of user-id/group-id testing along with the names. (I do think names are far more important, but if we are going to support uid/gid we should test it)

File: test/test_pebble.py
Comment (jameinel) [inline test/test_pebble.py:1371]: it seems odd to have a straight copy of this function here. Is it reasonable to have a separate _private/multipart.py file and import it from both places?

File: test/test_pebble.py
Comment (jameinel) [inline test/test_pebble.py:1509]: I didn't review all of these tests with a fine tooth comb. If there are any that you have concerns about, I'm happy to focus more on them.

File: test/test_pebble.py
Comment (jameinel) [inline test/test_pebble.py:1559]: I'm curious why you pulled 'dir_item' out and indirect it, vs including it in the content so that we can see the full request structure.

**Review by jameinel — COMMENTED** [review]
I do have concerns about the FileInfo using uid/gid instead of names, and whether we want read_file to take a BytesIO rather than returning one. (I suppose it is better for streaming and bulk, but it feels like a clumsy api from a charmer perspective)

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: Fixed.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py:615]: I don't think so -- in other places we just let the ValueError exception bubble up, which I think is the right thing to do. If Pebble is sending malformed data, something is seriously wrong (either a bug in our parsing code, or Pebble is really messed up).

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: Fixed per John -- just removed the quotes.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: It was copied from the `minBoundaryLength` in the server. However, per John, I've just avoided the length check altogether and only check whether it's present or not.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: Fair -- updated.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: Good call. Per Mattermost discussion, I've updated list-files in Pebble to return the user and group name too (cached per call so we don't normally have to do the lookup per entry). See https://github.com/canonical/pebble/pull/28

Comment (benhoyt) [reply ops/pebble.py]: Though also note that write_content and make_dir accept user/group either as UID/GID int or user/group name.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: Fixed.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: Yep, good call -- changed to ensure.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: Per discussion with Ian on Mattermost, it's actually a bit tricky to have this return a readable file-like object, because you're not yet finished reading the multipart response by the time `read_file` returns. So it makes the implementation tricky (you'd have to return some kind of object which reads bytes and then goes on to parse the metadata JSON in the last `read` call), and it makes error handling a bit weird, because protocol/path errors may be raised by the returned object's `read()` rather than the main `read_file()` call.

So what I've decided is to just include `read_content` / `write_content` for now, as those are the simplest APIs and we need those APIs anyway (and my implementation was reading everything into memory anyway). This will get us going, and we can always add streaming/bulk/whatever other APIs in a future release.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: Updated (at least the functions in this PR) to that style -- thanks.

File: test/test_pebble.py
Comment (benhoyt) [reply test/test_pebble.py:1397]: Added a test of both kinds - thanks.

File: test/test_pebble.py
Comment (benhoyt) [reply test/test_pebble.py]: It was somewhat different, and I didn't want to have to pollute the main code with test-specific options to try to unify it. In any case, I've removed this function in pebble.py now with the new read_content implementation.

File: test/test_pebble.py
Comment (benhoyt) [reply test/test_pebble.py:1616]: I don't have any concerns about them. They're very straight-forward as request and responses are mocked.

File: test/test_pebble.py
Comment (benhoyt) [reply test/test_pebble.py]: Rejiggered this a little.

**Comment by benhoyt** [comment]
One thing I'm not enthusiastic about is the copied docstrings in the model.Container versions of the APIs. We could just say `read_content.__doc__ = pebble.Client.read_content.__doc__`, but what our doc system (and RTD) handle that?

File: ops/pebble.py
Comment (niemeyer) [inline ops/pebble.py:878]: In previous conversations we tried to keep the API here consistent with the command line (the pebble command) and the API itself when that was possible. This sounds like a case we can use that idea again: we spent a while discussing the details of the method names, CLIs (push, pull, etc), so we might as well reuse these insights to produce an API here that will resemble what we have all around. It will make the muscle memory stronger.

File: ops/pebble.py
Comment (niemeyer) [inline ops/pebble.py:914]: Is there no support for streaming in the parser, or is it saving files in a temporary location similar to what Go does? We can't require the entire transference to be in memory. It's okay if we need to release v1 like that, but if that's what it takes we need to fix it shortly after.

File: ops/pebble.py
Comment (niemeyer) [inline ops/pebble.py:988]: Same question as above on the opposite end.

File: ops/pebble.py
Comment (niemeyer) [inline ops/pebble.py:1009]: What if we want to send/receive both a uid and a username? Seems safer to be explicit.

Same for groups below.

**Review by niemeyer — APPROVED** [review]
My Python knowledge is getting rusty with all the on-going changes in the language, but this looks pretty good in general. A few minor points below. LGTM once you're happy about these.

File: ops/pebble.py
Comment (jameinel) [reply ops/pebble.py:914]: He did have an underlying API that was reading into a BinaryIO object, but it lead to a clumsy interaction if you just want to see the content of a config file.
I think it is one that we could introduce once we have a nice pattern (should it be read_file(path, IO) or read_file(path) => IO)

I think we started with read_content because we know that we want to have a simple way to get small content, and we didn't want to commit to an API for streaming content until we were comfortable with how it hung together.

Comment (niemeyer) [reply ops/pebble.py:914]: How about `push(path, file_like)` and `pull(path) => file_like`?  `pull(path).read()` seems short and idiomatic.

Comment (benhoyt) [reply ops/pebble.py:914]: @niemeyer Yeah, that looks like a nice API at first glance, but 1) it's tricky to implement, as the file_like returned from `pull` would have to parse multipart in a streaming fashion -- there's no direct support for what we need in the stdlib -- and then do something on the last `file_like.read()` call to process the JSON metadata in the last multipart part, 2) because of that, it means that `pull` itself won't raise errors with the overall call, but the last `read()` may raise errors, which seems a bit weird to me.

On balance I do still think that's probably a reasonable API, but it'd take some time to implement correctly, so we would like to do it separately later. I can play with it some more if you feel strongly that we *don't* want `read_content` or `write_content`.

File: ops/pebble.py
Comment (jameinel) [reply ops/pebble.py:1009]: The issue for sending both uid and username is that I believe that creates ambiguity. Did you mean 1001 or user 'foo' who is 1002?

**Review by jameinel — APPROVED** [review]

File: ops/pebble.py
Comment (niemeyer) [reply ops/pebble.py:1009]: The answers to these specific questions seem easy: I meant user 'foo' with uid 1001, and 1002 is whoever 1002 is, of course. Those things are an inherent part of the underlying problem you are dealing with: there are _really_ two systems with their own data. So the more interesting question is what you do with that information.

For example, if we are transferring user 'foo' that has id 1001 to the local system, and neither foo nor 1001 exist locally and you want to create them, it's nice to replicate the exact data. If user 'foo' already exists, the better behavior seems to be preserving the username and warning that the uid diverges. We might also opt to forbid it entirely, but we can only do that if we have the data, and we might even get started by having both parameters and enforcing the use of a single one of them for now.

But if you have both uid and username under the same parameter, game over. The API was cooked in a way that you cannot even decide to forbid or catch mistakes, or evolve in a polished way, because we've encouraged people to use a lesser API upfront.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py:914]: After discussing with the team, we concluded we'll want string-returning and string-accepting functions anyway, so I'm going ahead with `read_content` and `write_content` for v1. We'll plan to add file-like object variants later -- likely named `push` and `pull`, once we settle on the right API and error handling.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: See below.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py:988]: See above.

File: ops/pebble.py
Comment (benhoyt) [reply ops/pebble.py]: If updated the `write_content` and `make_dir` functions signatures to accept both `user_id` and `user`, `group_id` and `group`. The current behavior (and this matches the underlying HTTP API) is if both are set, the IDs take precedence. Having both args also makes the types a bit clearer and the signature a bit more direct. The idea is people will use one or the other.

**Pull request merged**
